EQL: simplify equals/not-equals TRUE/FALSE expressions

astefan opened the pull request:
Simplify equals/not-equals TRUE/FALSE expressions, by transforming them into an AND(TRUE, new-expression) where new-expression may be the original expression or a negation of it. The aim is for the resulting expression to be simplified further by the subsequent optimization rules in the chain (AND(expression, TRUE) -> expression).

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/EQL)

astefan commented:
@elasticmachine update branch

elasticmachine commented:
user doesn't have permission to update head repository

rw-access commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -62,6 +62,57 @@ public Expression rule(Expression e) {
+                }
+                
+                if (newLeft != null) {
+                    return new And(e.source(), newLeft, newRight);
```
What's the purpose of the intermediate `And` node if it's always combined with `TRUE`?

rw-access commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java:
```diff
@@ -62,6 +62,57 @@ public Expression rule(Expression e) {
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
```
I think it's okay to assume that TRUE is on the right,  since we've moved literals to the right.
If it's on the left, than both left and right must be literals and the expression will fold anyway.

If that's true, I think it'll simplify the logic a good bit:
```java
if (e instanceof Equals ) {
    Equals eq = (Equals) e;
    if (TRUE.equals(eq.right()) {
        return eq.left()
    }

    if (FALSE.equals(eq.right()) {
        return new Not(eq.source(), eq.left())
    }
}

if (e instanceof NotEquals ) {
    Equals eq = (NotEquals) e;
    if (FALSE.equals(eq.right()) {
        return eq.left()
    }

    if (TRUE.equals(eq.right()) {
        return new Not(eq.source(), eq.left())
    }
}
```

costin replied:
`BooleanLiteralsOnTheRight` is applied after this rule so assuming literals on the right doesn't always work at least for the first iteration - I think this is fine or we could rearrange the rules, either move this rule after bools on the right or move that ahead.

I agree with Ross that the code is simplified this way and we could avoid the intermediate `And` as well (it would be simplified by another rule but there's no value in keeping it around).

costin replied:
Note you could compress the code even more though it impacts readability:
```java
if (e instanceof Equals || e instanceof NotEquals) {
    BinaryComparison bc = (BinaryComparison) e;
    Literal isTrue = e instanceof Equals ? TRUE : FALSE;
    Literal isFalse = e instanceof NotEquals ? TRUE: FALSE;

    // e == TRUE or e != FALSE
    if (isTrue.equals(bc.right()) {
        return bc.left();
    }

    //  e == FALSE or e != TRUE
    if (isFalse.equals(bc.right()) {
        return new Not(bc.source(), bc.left())
    }
}
```

astefan replied:
I can change this.
In the way it is now, this is a "standalone" rule that can be applied no matter what rules are after or before it. Maybe it can be reused in the future in other languages. Changing it will assume that `BooleanLiteralsOnTheRight` **must** be always before this rule. And the rule will be much simplified, as you pointed out.

astefan commented:
@costin @rw-access I've addressed the comments but, as [I mentioned](https://github.com/elastic/elasticsearch/pull/56191#discussion_r420607612), I think there was a good reason behind the original version.

astefan commented:
@elasticmachine update branch

elasticmachine commented:
user doesn't have permission to update head repository

astefan commented:
@elasticmachine update branch

elasticmachine commented:
user doesn't have permission to update head repository
